File: code/backend/app/schemas/plat.py
"""平台模块 Schema（审计日志响应等）。"""
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict

from app.schemas.common import BaseOut


class AuditLogOut(BaseOut):
    actor_id: Optional[int] = None
    role: Optional[str] = None
    action: Optional[str] = None
    resource: Optional[str] = None
    before_json: Optional[dict] = None
    after_json: Optional[dict] = None
    ip: Optional[str] = None
    # P4 哈希链防篡改字段
    seq_no: int = 0
    prev_hash: Optional[str] = None
    hash: Optional[str] = None


class PredLogOut(BaseOut):
    """AI 预测日志输出（第15.4章 反馈闭环）。"""

    model_id: Optional[int] = None
    version: Optional[str] = None
    customer_id: Optional[int] = None
    input_features_json: Optional[dict] = None
    predict_result_json: Optional[dict] = None
    predict_time: Optional[datetime] = None
    user_id: Optional[int] = None
    adopted: Optional[str] = None


# ---- AI 模型目录（plat_ai_model，第11.2/12.3章） ----


class AiModelCreate(BaseModel):
    name: str
    version: str
    train_dataset_id: Optional[int] = None
    algo_type: Optional[str] = None
    metrics_json: Optional[dict] = None
    status: str = "offline"


class AiModelUpdate(BaseModel):
    name: Optional[str] = None
    version: Optional[str] = None
    train_dataset_id: Optional[int] = None
    algo_type: Optional[str] = None
    metrics_json: Optional[dict] = None
    status: Optional[str] = None


class AiModelOut(BaseOut):
    name: Optional[str] = None
    version: Optional[str] = None
    train_dataset_id: Optional[int] = None
    algo_type: Optional[str] = None
    metrics_json: Optional[dict] = None
    status: str = "offline"
    online_at: Optional[datetime] = None
    offline_at: Optional[datetime] = None


# ---- 数据资产目录（plat_data_asset，第11.2/13章） ----


class DataAssetCreate(BaseModel):
    name: str
    owner: Optional[str] = None
    sensitivity_level: Optional[str] = None  # L1-L4
    usage_scope: Optional[str] = None
    quality_score: Optional[float] = None
    update_freq: Optional[str] = None
    lineage_json: Optional[dict] = None


class DataAssetUpdate(BaseModel):
    name: Optional[str] = None
    owner: Optional[str] = None
    sensitivity_level: Optional[str] = None
    usage_scope: Optional[str] = None
    quality_score: Optional[float] = None
    update_freq: Optional[str] = None
    lineage_json: Optional[dict] = None


class DataAssetOut(BaseOut):
    name: Optional[str] = None
    owner: Optional[str] = None
    sensitivity_level: Optional[str] = None
    usage_scope: Optional[str] = None
    quality_score: Optional[float] = None
    update_freq: Optional[str] = None
    lineage_json: Optional[dict] = None


# ---------------- P6: 合规采集审核 ----------------
class ComplianceSubmitIn(BaseModel):
    category: str  # qualification/license/privacy/security/...
    subject_type: str  # doctor/store/therapist/platform
    subject_id: Optional[int] = None
    title: str
    content_json: Optional[dict] = None


class ComplianceReviewIn(BaseModel):
    review_note: Optional[str] = None


class ComplianceOut(BaseOut):
    category: str
    subject_type: str
    subject_id: Optional[int] = None
    title: str
    content_json: Optional[dict] = None
    submitter_id: Optional[int] = None
    status: str
    reviewer_id: Optional[int] = None
    review_note: Optional[str] = None
    reviewed_at: Optional[datetime] = None

File: code/admin-web/src/services/plat.ts
// 平台管理（plat）服务封装（第10.2/11.2章）：AI 模型目录 / 数据资产目录。
// 统一响应经 request.ts 拦截器解包为 data（PageResult / 实体）。
import http from './request'
import { API } from '@/constants/api'

export interface PageResult<T> {
  total: number
  page: number
  page_size: number
  items: T[]
}

export interface AiModel {
  id: number
  name: string
  version: string
  algo_type?: string
  metrics_json?: Record<string, unknown>
  status: string
  online_at?: string
  offline_at?: string
  created_at?: string
  updated_at?: string
}

export interface DataAsset {
  id: number
  name: string
  owner?: string
  sensitivity_level?: string
  usage_scope?: string
  quality_score?: number
  update_freq?: string
  lineage_json?: Record<string, unknown>
  created_at?: string
  updated_at?: string
}

export async function listAiModels(params: {
  page?: number
  page_size?: number
  name?: string
  status?: string
  algo_type?: string
}) {
  return http.get(API.aiModels, { params }) as Promise<PageResult<AiModel>>
}

export async function listDataAssets(params: {
  page?: number
  page_size?: number
  name?: string
  owner?: string
  sensitivity_level?: string
}) {
  return http.get(API.dataAssets, { params }) as Promise<PageResult<DataAsset>>
}

// ---------------- 模型上下线 / 删除 ----------------
export async function setModelOnline(modelId: number) {
  return http.post(`${API.aiModels}/${modelId}/online`) as Promise<AiModel>
}

export async function setModelOffline(modelId: number) {
  return http.post(`${API.aiModels}/${modelId}/offline`) as Promise<AiModel>
}

export async function deleteModel(modelId: number) {
  return http.delete(`${API.aiModels}/${modelId}`) as Promise<void>
}

export async function createDataAsset(body: Partial<DataAsset>) {
  return http.post(API.dataAssets, body) as Promise<DataAsset>
}

export async function updateDataAsset(assetId: number, body: Partial<DataAsset>) {
  return http.put(`${API.dataAssets}/${assetId}`, body) as Promise<DataAsset>
}

export async function deleteDataAsset(assetId: number) {
  return http.delete(`${API.dataAssets}/${assetId}`) as Promise<void>
}

// ---------------- 审计日志（3.6.1 合规大脑，复用 P4 哈希链） ----------------
export interface AuditLog {
  id: number
  actor_id?: number
  role?: string
  action?: string
  resource?: string
  before_json?: Record<string, unknown>
  after_json?: Record<string, unknown>
  ip?: string
  seq_no: number
  prev_hash?: string
  hash?: string
  created_at?: string
}

export async function listAuditLogs(params: {
  page?: number
  page_size?: number
  resource?: string
}) {
  return http.get(API.auditLogs, { params }) as Promise<PageResult<AuditLog>>
}

export async function verifyAuditChain() {
  return http.get(`${API.auditLogs}/verify`) as Promise<{
    ok: boolean
    broken_at_seq: number | null
  }>
}

// ---------------- P6: 合规采集审核（3.6.2 合规大脑·采集与审核） ----------------
export interface ComplianceItem {
  id: number
  category: string
  subject_type: string
  subject_id?: number
  title: string
  content_json?: Record<string, unknown>
  submitter_id?: number
  status: string
  reviewer_id?: number
  review_note?: string
  reviewed_at?: string
  created_at?: string
}

export async function submitCompliance(body: {
  category: string
  subject_type: string
  subject_id?: number
  title: string
  content_json?: Record<string, unknown>
}) {
  return http.post(`${API.compliance}/submit`, body) as Promise<ComplianceItem>
}

export async function listCompliance(params: {
  category?: string
  status?: string
  subject_type?: string
  page?: number
  page_size?: number
}) {
  return http.get(API.compliance, { params }) as Promise<PageResult<ComplianceItem>>
}

export async function getCompliance(itemId: number) {
  return http.get(`${API.compliance}/${itemId}`) as Promise<ComplianceItem>
}

export async function approveCompliance(itemId: number, review_note?: string) {
  return http.post(`${API.compliance}/${itemId}/approve`, { review_note }) as Promise<ComplianceItem>
}

export async function rejectCompliance(itemId: number, review_note: string) {
  return http.post(`${API.compliance}/${itemId}/reject`, { review_note }) as Promise<ComplianceItem>
}
